eclair-core/src/main/scala/fr/acinq/eclair/router/BalanceEstimate.scala

```diff
@@ -0,0 +1,182 @@
+    baseline * (1 - decayRatio) + successProbabilityAtT * decayRatio
+  }
+
+  def couldNotSend(amount: MilliSatoshi, timestamp: TimestampSecond): BalanceEstimate =
+    if (amount < high) {
+      if (amount > low) {
+        copy(high = amount, highTimestamp = timestamp)
+      } else {
+        // the balance is actually below `low`, we discard our lower bound
+        copy(low = MilliSatoshi(0), lowTimestamp = timestamp, high = amount, highTimestamp = timestamp)
+      }
+    } else {
+      /* We already expected not to be able to relay that amount as it above our upper bound. However if the upper bound
+       * was old enough that replacing it with the current amount decreases the success probability for `high`, then we
+       * replace it.
+       */
+      val pLow = decay(low, 1, lowTimestamp)
+      val pHigh = decay(high, 0, highTimestamp)
+      val x = low + (high - low) * (pLow / (pLow - pHigh))
+      if (amount <= x) {
+        copy(high = amount, highTimestamp = timestamp)
+      } else {
+        copy()
+      }
+    }
+
+  def couldSend(amount: MilliSatoshi, timestamp: TimestampSecond): BalanceEstimate =
+    otherSide.couldNotSend(totalCapacity - amount, timestamp).otherSide
+
+  def didSend(amount: MilliSatoshi, timestamp: TimestampSecond): BalanceEstimate = {
+    val newLow = (low - amount) max MilliSatoshi(0)
+    val newHigh = (high - amount) max MilliSatoshi(0)
+    val pLow = decay(newLow, 1, lowTimestamp)
+    val pHigh = decay(newHigh, 0, highTimestamp)
+    if (???) {
+      copy(low = newLow, high = (totalCapacity - amount) max MilliSatoshi(0), highTimestamp = timestamp)
+    } else {
+      copy(low = newLow, high = newHigh)
+    }
+  }
+
+  def addChannel(capacity: Satoshi): BalanceEstimate = copy(high = high + toMilliSatoshi(capacity), totalCapacity = totalCapacity + capacity)
+
+  def removeChannel(capacity: Satoshi): BalanceEstimate = copy(low = (low - toMilliSatoshi(capacity)) max MilliSatoshi(0), high = high min toMilliSatoshi(totalCapacity - capacity), totalCapacity = totalCapacity - capacity)
+
+  /* Estimate the probability that we can successfully send `amount` through the channel
+   *
+   * We estimate this probability with a piecewise linear function:
+   * - probability that it can relay a payment of 0 is 1
+   * - probability that it can relay a payment of low is close to 1 if lowTimestamp is recent
+   * - probability that it can relay a payment of high is close to 0 if highTimestamp is recent
+   * - probability that it can relay a payment of totalCapacity is 0
+   */
+  def canSend(amount: MilliSatoshi): Double = {
+    val x = amount.toLong.toDouble
+    val a = low.toLong.toDouble
+    val b = high.toLong.toDouble
+    val c = toMilliSatoshi(totalCapacity).toLong.toDouble
+
+    // Success probability at the low and high points
+    val pLow = decay(low, 1, lowTimestamp)
+    val pHigh = decay(high, 0, highTimestamp)
+
+    if (amount < low) {
```

Comment:
In this case the probability of sending the amount should be 1 as you are certain that you can send the amount!

However taking your notation `a-x` is the amount that can be sent. However I do not understand the semantics of x*pLow and what is acchieved by dividing all by the lower bound of the channel `a ` It certainly does not give you the satoshis that may savely be sent and it most certaintly does not give you an accurate probability. 

I have similar issues understanding the next 5 lines.

Reply:
No.
`low` was the lower bound at some time t. If you call `canSend` at that time t, then yes the result will be 1 but if call it later it will be lower than 1.